src/Grpc.Net.Client/Balancer/Subchannel.cs

```diff
@@ -356,7 +378,11 @@ public IReadOnlyList<BalancerAddress> GetAddresses()
         public void Dispose()
         {
             UpdateConnectivityState(ConnectivityState.Shutdown, "Subchannel disposed.");
-            _stateChangedRegistrations.Clear();
+            for (var i = _stateChangedRegistrations.Count - 1; i >= 0; i--)
+            {
+                _stateChangedRegistrations[i].Dispose();
```

Comment:
This is removing itself from the list that you are currently trying to iterate over, that is not good.

Reply:
The for loop is reversed so it doesn't error.

Do you think a `while (_stateChangedRegistrations.Count > 0)` would be better?

Reply:
Oh man, that's confusing. Does the remove really need to be separate? Some form of rewrite would make me happier :D

Reply:
I want to use dispose here to remove the registrations because it has logging. It means removing registrations is always consistent (dispose) vs something inconsistent (dispose vs Clear).